File: p022m/generate_parenthesis.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function (n) {
  const ans = [];
  const backtrack = (s, open, close, max) => {
    // console.log("s =", s, ", open =", open, ", close =", close);
    if (s.length === max * 2) {
      // console.log("Ouptut:", s);
      ans.push(s);
      return;
    }
    if (open < max) backtrack(s + '(', open + 1, close, max);
    if (open > close) backtrack(s + ')', open, close + 1, max);
  };

  backtrack('', 0, 0, n);
  return ans;
};

var argv2 = process.argv[2];
var n = argv2 === undefined ? 0 : parseInt(argv2);
console.log(generateParenthesis(n));

File: p022m/generate_parenthesis.cpp
// Build program by
// g++ -std=c++11 *.cpp -o test && ./test && rm test
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;
class Solution {
public:
    vector<string> generateParenthesis(int n)
    {
        vector<string> ans;
        std::function<void(string, int, int)> backtrack =
            [&](string ps, int open, int close) -> void {
            if (open == 0 && close == 0) {
                ans.push_back(ps);
                return;
            }
            if (open > 0)
                backtrack(ps + "(", open - 1, close + 1);
            if (close > 0)
                backtrack(ps + ")", open, close - 1);
        };
        backtrack("", n, 0);
        return ans;
    }
};
// TEST
int main(int argc, char* argv[])
{
    if (argc <= 1) {
        cout << "Please input a number" << endl;
        return 1;
    }
    stringstream ss(argv[1]);
    int n = 0;
    ss >> n;
    Solution sol;
    vector<string> ans = sol.generateParenthesis(n);
    for (const string& s : ans) {
        cout << s << endl;
    }
    return 0;
}
